tsdb/wlog/watcher.go

```diff
@@ -502,24 +487,73 @@ func (w *Watcher) readSegment(r *LiveReader, segmentNum int, tail bool) error {
 				return err
 			}
 			w.writer.StoreSeries(series, segmentNum)
+		case record.Unknown:
+			// Could be corruption, or reading from a WAL from a newer Prometheus.
+			w.recordDecodeFailsMetric.Inc()
+		default:
+			// We're not interested in other types of records.
+		}
+	}
+	if err := r.Err(); err != nil && !errors.Is(err, io.EOF) {
+		return fmt.Errorf("segment %d: %w", segmentNum, err)
+	}
+	return nil
+}
+
+// readSegment reads all known records into w.writer from a segment.
+func (w *Watcher) readSegment(r *LiveReader, startT int64, segmentNum int) (err error) {
+	var (
+		// One table per WAL segment means it won't grow indefinitely.
+		dec = record.NewDecoder(labels.NewSymbolTable())
+
+		series                = w.seriesPool.Get()[:0]
+		samples               = w.samplesPool.Get()[:0]
+		samplesToSend         = w.samplesPool.Get()[:0]
+		exemplars             = w.exemplarsPool.Get()[:0]
+		histograms            = w.histogramsPool.Get()[:0]
+		histogramsToSend      = w.histogramsPool.Get()[:0]
+		floatHistograms       = w.floatHistogramsPool.Get()[:0]
+		floatHistogramsToSend = w.floatHistogramsPool.Get()[:0]
+		metadata              = w.metadataPool.Get()[:0]
+	)
+
+	defer func() {
+		// WriteTo interface explicitly calls out those slices to be immutable.
+		w.seriesPool.Put(series)
+		w.samplesPool.Put(samples)
+		w.samplesPool.Put(samplesToSend)
+		w.exemplarsPool.Put(exemplars)
+		w.histogramsPool.Put(histograms)
+		w.histogramsPool.Put(histogramsToSend)
+		w.floatHistogramsPool.Put(floatHistograms)
+		w.floatHistogramsPool.Put(floatHistogramsToSend)
+		w.metadataPool.Put(metadata)
+	}()
+
+	for r.Next() && !isClosed(w.quit) {
+		rec := r.Record()
+		w.recordsReadMetric.WithLabelValues(dec.Type(rec).String()).Inc()
+
+		switch dec.Type(rec) {
+		case record.Series:
+			series, err = dec.Series(rec, series[:0])
+			if err != nil {
+				w.recordDecodeFailsMetric.Inc()
+				return err
+			}
+			w.writer.StoreSeries(series, segmentNum)
 
 		case record.Samples:
-			// If we're not tailing a segment we can ignore any samples records we see.
-			// This speeds up replay of the WAL by > 10x.
-			if !tail {
-				break
-			}
-			samples, err := dec.Samples(rec, samples[:0])
+			samples, err = dec.Samples(rec, samples[:0])
```

Comment:
You don't think we need the skip-samples optimisation any more?

Reply:
I'm 100% sure we don't need, because we already use `readSegmentSeries` for tailing which ignored all but series (in my PR and before my PR). This `!tail` if was never utilized in our prod code - that's the one of many reasons to unconfuse this code.. (:

Reply:
This is confusing to me; I see `readSegmentSeries` as added in this PR.
Note I did have my own attempt at rewriting https://github.com/prometheus/prometheus/pull/14439/files.

Reply:
Ups, you are right, sorry for confusion -- edited for what's true then. I just found this weird `tail` bool confusing, trying to reduce the use of it.

> Note I did have my own attempt at rewriting https://github.com/prometheus/prometheus/pull/14439/files.

TIL, will check if there is something I could learn from your attempt, sorry for missing!